Replace Append-only with Verifiable Data Structure (#351)
ietf-wg-scitt/draft-ietf-scitt-architecture: draft-ietf-scitt-architecture.md

    target: https://in-toto.io/
    title: in-toto

>   MERKLE: DOI.10.1007/3-540-48184-2_32

Reviewer: If we are deleting the references to Merkle trees, which are likely foundational, and we will start talking about VDS, should we not explain where that notion comes from? It is built up in IETF specs and standards, but we do mean to cite the origin of the idea?

```suggestion
 VDS: https://github.com/google/trillian/blob/master/docs/papers/VerifiableDataStructures.pdf
```
That said, I am not attached to this change and think this path is not constructive for the I-D.

Author: @aj-stein  We just saw this. Can you please open a new PR?

Reviewer 2: vds is defined in cose receipts

The term "claim" is defined in {{RFC8392}}.

> Append-only Log:
> 
> A sequence of Signed Statements captured by a Verifiable Data Structure.

Reviewer 3: ```suggestion
Registered Sequence of Signed Statements:

A sequence of Signed Statements submitted by Clients that meet applicable registration policies and are captured by a Verifiable Data Structure.
```

Proposal for a term to designate the sequence of signed statements committed to the VDS, as opposed to the VDS itself, and hopefully move away from the controversial "Append-only Log". It's useful to have a term for this, when discussing audit.

Reviewer 2: Append-only VDS?

Author: Resolving, with Issue #352

The term "claim" is defined in {{RFC8392}}.

Append-only Log:

> : see Verifiable Data Structure

Reviewer 3: "A sequence of entries captured by a Verifiable Data Structure"? The log isn't itself a VDS, right?

Author: @achamayou, Could you make a suggestion?

> : see Verifiable Data Structure

Reviewer 3: ```suggestion
A sequence of Signed Statements captured by a Verifiable Data Structure.

: see Verifiable Data Structure
```

Equivocation:

> : a state where it is possible for a Transparency Service to provide different views of its Verifiable Data Structure to Relying Parties about the same Artifact {{EQUIVOCATION}}.

Reviewer 3: ```suggestion
: a state where it is possible for a Transparency Service to provide different Verifiable Data Structures to Relying Parties about the same Artifact {{EQUIVOCATION}}.
```

Reviewer 3: I think this should be Signed Statement rather than Artifact?

Equivocation:

> : a state where it is possible for a Transparency Service to provide different Verifiable Data Structures to Relying Parties about the same Artifact {{EQUIVOCATION}}.

Reviewer 3: ```suggestion
: a state where a Transparency Service provides incompatible Verifiable Data Proofs to Relying Parties, containing conflicting claims about the Signed Statement bound at a given position in the Verifiable Data Structure {{EQUIVOCATION}}.
```

The service wouldn't provide the whole VDS in general, only VDPs/proofs extracted from it, so I think the practical manifestation of equivocation for observers is the presence of incompatible VPDs (which are certainly evidence of incompatible VDSs existing somewhere, but which the service will not usually provide wholesale).

Reviewer 4: ```suggestion
: a state where a Transparency Service provides inconsistent proofs to Relying Parties, containing conflicting claims about the Signed Statement bound at a given position in the Verifiable Data Structure {{EQUIVOCATION}}.
```
Maybe avoid introducing Verifiable Data Proofs ? I think we can reasonably just say "proofs" if the context is a Verifiable Data Structure

Also a nit: I think inconsistent is more accurate than "incompatible" especially given the other contexts interoperability and compatibility are used

Non-equivocation:

> : a state where it is impossible for a Transparency Service to provide different views of its Verifiable Data Structure to Relying Parties about the same Artifact.

Reviewer 3: ```suggestion
: a state where it is impossible for a Transparency Service to provide different Verifiable Data Structures to Relying Parties about the same Artifact.
```

Over time, an Issuer may register new Signed Statements about an Artifact in a Transparency Service with new information.
However, the consistency of a collection of Signed Statements about the Artifact can be checked by all Relying Parties.

Receipt:

> : a cryptographic proof that a Signed Statement is included in the Verifiable Data Structure.

Author: ```suggestion
: a cryptographic proof that a Signed Statement is included in the Verifiable Data Structure.
See {{-COMETRE}} for implementations

Non-equivocation:

> : a state where it is impossible for a Transparency Service to provide different Verifiable Data Structures to Relying Parties about the same Artifact.

Reviewer 3: ```suggestion
: a state where all Verifiable Data Proofs provided by the Transparency Service to Relying Parties are produced from a Single Verifiable Data Structure describing a unique sequence of Signed Statements, and are therefore compatible.
```

Similar to the clarification above, probably a bit heavy handed, but trying to avoid stating that the VDS is provided to RPs.

Reviewer 4: ```suggestion
: a state where all proofs provided by the Transparency Service to Relying Parties are produced from a Single Verifiable Data Structure describing a unique sequence of Signed Statements, and are therefore consistent.
```

avoid adding "Verifiable Data Proofs" and just rely on mention of VDS for context ?
Also the nit: compatible/consistent

Transparency Service:

: an entity that maintains and extends the Verifiable Data Structure, and endorses its state.

> A Transparency Service can be a complex system, requiring the Transparency Service to provide many security guarantees about its Append-only Log.

Reviewer 3: ```suggestion
```

I was going to do edit this sentence, but I think it's actually almost information-free, and so I propose we drop it.

Transparency Service:

> : an entity that maintains and extends the Verifiable Data Structure, and endorses its state.

Reviewer 3: ```suggestion
: an entity that maintains and extends the Append-only Log, a Verifiable Data Structure that captures, and endorses it.
```

I think we still intend to have the TS maintain the log itself, as well as the VDS.

Reviewer 5: Not sure about this, lets discuss

Author: We really need to decide if we're going to define the Append-only Log/ Verifiable Data Structure as two distinct things.

Author: deferring to #352

> A Transparency Service can be a complex system, requiring the Transparency Service to provide many security guarantees about its Verifiable Data Structure.

Reviewer 3: ```suggestion
A Transparency Service can be a complex system, requiring the Transparency Service to provide many security guarantees about its Append-only Log.
```

I think the useful end guarantee is about the log and its entries, the VDS is the mechanism used to implement the guarantees.

SCITT supports multiple Verifiable Data Structures and Receipt formats as defined in {{-COMETRE}}, accommodating different Transparency Service implementations.
May be referenced as an Append-only log, indicating it's immutable nature.
{: #mybody}

# Definition of Transparency

In this document, the definition of transparency is intended to build over abstract notions of Append-only Logs and Receipts.
Existing transparency systems such as Certificate Transparency are instances of this definition.
SCITT supports multiple Append-only log formats, as defined in {{-COMETRE}}, and uses Verifiable Data Structure throughout the document.

A Signed Statement is an identifiable and non-repudiable Statement made by an Issuer.
The Issuer selects additional metadata and attaches a proof of endorsement (in most cases, a signature) using the identity key of the Issuer that binds the Statement and its metadata.
Signed Statements can be made transparent by attaching a proof of Registration by a Transparency Service, in the form of a Receipt.

> Receipts demonstrate inclusion of Signed Statements in the Verifiable Data Structure of a Transparency Service.

Reviewer 3: ```suggestion
Receipts demonstrate inclusion of Signed Statements in the Append-only Log of a Transparency Service.
```

Reviewer 5: I don't like this.

Author: Can you elaborate on which you don't like? The PR change or the Suggestion?

Reviewer 5: Because I don't know what an "append only log is"... if it is a normative dependency, and it is not a VDS, what are its requirements beyond the requirements of a VDS?

Reviewer 3: There's a logical sequence of statements, which is not a normative dependency but is important when discussing audit. I think we can keep this line that way though.

Verifiable Data Structure:

: a data structure which supports one or more proof types, such as "inclusion proofs" or "consistency proofs", for Signed Statements as they are Registered to a Transparency Service.
SCITT supports multiple Verifiable Data Structures and Receipt formats as defined in {{-COMETRE}}, accommodating different Transparency Service implementations.

> May be referenced as an Append-only log, indicating its immutable nature.

Reviewer 3: ```suggestion
```

The log is the sequence of full entries, the VDS captures their digest and can be used to emit VDPs, but is not the same thing as the log itself.

Reviewer 4: Yeah this one change I think captures my earlier point. A VDS proves append only, the storage can not do this. So it is reasonable to say "the vds assures an append only log", and maybe also shorthand that as "referenced as an  Append-only log"

Author: Leaving this open for discussion as I'm not hearing convergence to remove, or alter the line.

Reviewer 3: My preference would be to drop the line, and clearly distinguish the "Register Sequence of Signed Statements" (*) and the "VDS" elsewhere too. (*) I am not hung up on the particularly name, but something that describes the Signed Statements themselves in the order in which they are committed to the VDS.

{: #mybody}

# Definition of Transparency

In this document, the definition of transparency is intended to build over abstract notions of Append-only Logs and Receipts.
Existing transparency systems such as Certificate Transparency are instances of this definition.

> SCITT supports multiple Append-only log formats, as defined in {{-COMETRE}}, and uses Verifiable Data Structure throughout the document.

Reviewer 3: ```suggestion
SCITT supports multiple Verifiable Data Structures, as defined in {{-COMETRE}}.
```

Transparency is implemented by providing a consistent, append-only, cryptographically verifiable, publicly available record of entries.
A SCITT instance is referred to as a Transparency Service.
Implementations of Transparency Services may protect their Append-only Log and Verifiable Data Structure using a combination of trusted hardware, consensus protocols, and cryptographic evidence.

> A Receipt is an offline, universally-verifiable proof that an entry is registered in the Verifiable Data Structure.

Reviewer 3: ```suggestion
A Receipt is a signature over one or more Verifiable Data Structure Proofs that an entry is registered in the Verifiable Data Structure. It is universally verifiable without online access to the TS.
```

Split for clarity?

Reviewer 4: I particularly like "It is universally verifiable without online access to the TS"

Reviewer 4: That has a stronger privacy posture than CT I think

Author: I'm committing the suggestion, we can review in totality.

Transparency is implemented by providing a consistent, append-only, cryptographically verifiable, publicly available record of entries.
A SCITT instance is referred to as a Transparency Service.

> Implementations of Transparency Services may protect their Append-only Log and Verifiable Data Structure using a combination of trusted hardware, consensus protocols, and cryptographic evidence.

Reviewer 3: ```suggestion
Implementations of Transparency Services may protect their Registered Sequence of Signed Statements and Verifiable Data Structure using a combination of trusted hardware, consensus protocols, and cryptographic evidence.
```

> A Receipt is a signature over one or more Verifiable Data Structure Proofs that an entry is registered in the Verifiable Data Structure.

Reviewer 3: ```suggestion
A Receipt is a signature over one or more Verifiable Data Structure Proofs that a Signed Statement is registered in the Verifiable Data Structure.
```

Transparency is implemented by providing a consistent, append-only, cryptographically verifiable, publicly available record of entries.
A SCITT instance is referred to as a Transparency Service.

> Implementations of Transparency Services may protect their Verifiable Data Structure using a combination of trusted hardware, replication and consensus protocols, and cryptographic evidence.

Reviewer 3: ```suggestion
Implementations of Transparency Services may protect their Append-only Log and Verifiable Data Structure using a combination of trusted hardware, consensus protocols, and cryptographic evidence.
```

Replication is a kind of consensus, TS implementation that provide for example the ability to fetch Transparent Statements and not just receipts will want to protect the whole log, not just a VDS. Even implementations that don't may simply derive the VDS from a protected log.

Reviewer 5: I don't like this change, it confuses the security mechanism for receipts, with securing mechanisms for storing VDS.

Author: @OR13, can you clarify if you support the current PR, or the suggested change?

Transparency is implemented by providing a consistent, append-only, cryptographically verifiable, publicly available record of entries.
A SCITT instance is referred to as a Transparency Service.
Implementations of Transparency Services may protect their Append-only Log and Verifiable Data Structure using a combination of trusted hardware, consensus protocols, and cryptographic evidence.

> A Receipt is an offline, universally-verifiable proof that an entry is registered in the Verifiable Data Structure.

Reviewer 3: Is Receipt just an alias for a VDP as defined in COMETRE? Should we update the definition accordingly and drop this line altogether?

Reviewer 4: Is a VDP defined in CoMETER? I cant find Verifiable Data Proof here https://github.com/ietf-scitt/draft-birkholz-cose-cometre-ccf-profile/blob/main/draft-birkholz-cose-receipts-ccf-profile.md

I think Receipt is not an alias of VDP. It is a container for proof material that is verified against a VDS and is described by a profile of https://cose-wg.github.io/draft-ietf-cose-merkle-tree-proofs/draft-ietf-cose-merkle-tree-proofs.html such as COMETER. Should this be a reference to that instead ?

Reviewer 4: agree. we already utilise the multiple proofs so its not just a theoritical thing. we found it actively useful

The protected header of the COSE_Sign1 Envelope MUST include either the Issuer's certificate as `x5t` or the chain including the Issuer's certificate as `x5chain`.
If `x5t` is included in the protected header, an `x5chain` with a leaf certificate corresponding to the `x5t` value MAY be included in the unprotected header.

> The Transparency Service MUST apply the Registration Policy that was most recently added to the Verifiable Data Structure at the time of Registration.

Reviewer 3: ```suggestion
The Transparency Service MUST apply the Registration Policy that was most recently added to the Append-only Log at the time of Registration.
```

The VDS records only digests of entries in the log, it will contain a commitment to, but not the policy itself.

Reviewer 5: Merkle trees only store digests... other stuff might store the full signed statement.
Don't assume merkle trees would be my recommendation.
The important part is that the client accepts the security mechanism used by the TS... submission to a TS implies the desire to have a receipt produced from the VDS... regardless of if the VDS is a log or not.

Reviewer 5: If we want to restrict the architecture to specific VDS (CT / CCF logs), that seems like a better solution than using the word "append only log".

Reviewer 3: I've re-written the suggestion to use VDS, but use the word "commit" rather than "add". The VDS produces commitments to entries, but not entries themselves.

I don't know what the other stuff is, CT, CCF and the MMR all use Merkle Trees. The only "example" ever mentioned in the COMETRE spec is Merkle Trees also. I am always happy to learn, but at the moment I do not believe that confusing the VDS and the log of full entries is net positive.

Reviewer 4: @achamayou there exist production systems which use mathematical objects, rather than hashes, for the commitment scheme. "trees of math" can be built out of these just fine and can offer advantages in size and privacy. 

A more readily comparible approach to merkle trees is this one - https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf 

The minima blockchain uses a clever recursive proof to provide a constant size commitment to the entire state of their chain.

There is quite a lot of overlap with the verifiable compute and zk spaces. Nova and Halo. See this eprint for a starting example. https://eprint.iacr.org/2023/969.pdf 

The intent has always been to avoid excluding these sorts of things, and we prototyped verkle's in 2023.

#### Auditability of Registration

The operator of a Transparency Service MAY update the Registration Policy or the trust anchors of a Transparency Service at any time.

> Transparency Services MUST ensure that for any Signed Statement they register, enough information is made available to Auditors (either in the Verifiable Data Structure and retrievable through audit APIs, or included in the Receipt) to reproduce the Registration checks that were defined by the Registration Policies at the time of Registration.

Reviewer 3: ```suggestion
Transparency Services MUST ensure that for any Signed Statement they register, enough information is made available to Auditors (either in the Append-only Log and retrievable through audit APIs, or included in the Receipt) to reproduce the Registration checks that were defined by the Registration Policies at the time of Registration.
```

The VDS is just a tree of digests in most cases, it does not directly contain the information required for audit, only commitments to that information. The log is what contains the information.

Reviewer 5: yeah, I agree the sentence is problematic.

using the word append only log does not improve it though.

An auditor needs access to adjacent storage and the VDS to perform certain security related checks (like proving correct construction).

Author: > The log is what contains the information.

What information is on each element? For the average reader, it's not clear whether the contents of the signed statement are included in the log, within the VDS, or a cryptographic proof of the signed statement is recorded.
Related to the PII discussion, if someone stores a name or a phone number in the payload or in a header, what would it take to "redact" that information?
If the log only stores hashes or other cryptographic proof of the signed statement, which contains a payload, then there's nothing to actually redact.
If the TS stores the signed statements in an ancillary service, they can delete the signed statement from that external storage, and the log/vds remains intact.

The purpose of defining what actually goes into the log was meant to clarify this.

Reviewer 3: > If the TS stores the signed statements in an ancillary service, they can delete the signed statement from that external storage, and the log/vds remains intact.

Depending on how registration policies/trust anchors are recorded, this potentially affects auditability.

> the log/vds remains intact.

Maybe this is a terminology issue, I see the log as the sequence of Signed Statements, and the VDS as the datastructure over their identities/digests. I don't want to get stuck on this, I am happy to talk about VDSs, and to use another name for the sequence of Signed Statements that's committed in the VDS.

If I am an auditor, and I am trying to figure out if a set of entries were legitimately registered, I need to be able to get the applicable trust anchors/policies at the time. I can't do that from a TS that only has a Merkle Tree. I think this is really important, it's not just a secondary concern, I am not happy with a TS that produces a consistent but arbitrary sequence full of unauthenticated/incorrectly registered entries.
A TS that can arbitrarily elide the contents of policy-updating statements can trivially hide mis-behaviour, and claim to have been enforcing an older policy for example.

Anyway, that's a long winded way to explain why I think it's important that the full sequence of entries/log is not completely written out and replaced by the VDS :)

Reviewer 4: > Anyway, that's a long winded way to explain why I think it's important that the full sequence of entries/log is not completely written out and replaced by the VDS :)

Maybe this is a terminology thing. We took the view that "append only" must be verifiable and therefore the ability to prove this is intrinsic to the VDS. Having *both* append only and also a VDS is causing even more confusion because append only is just a property that is provable by the VDS (imo).

We also took the view that provability of position and sequence also must be verifiable and so required this of our VDS.

My perspective is that what we are discussing here as "append only log" is just a database. Its not "append only" in any verifiable sense on its own. Even if the database tech is some storage system that enforces append only on clients, this cant be proven without a suitable VDS.

I think I'm back to the original intent of this pr which I believe was to first replace "Append Only" entirely with VDS and then take the issue of clarifying how the storage of statements during registration relates to the vds as a separate issue ?

Reviewer 3: @robinbryce this helps a lot, what I would like not to lose here is "Log" as in "Sequence of Signed Statements", I am happy to see "Append-only" go. It's not confusing to me (I think we implement a VDS to make it effectively Append-only, or more modestly "tamper-evident" as you say), but it's also not necessary so long as it's clarified elsewhere (which it is, quite a bit).

I am hesitant to start discussing whether it is possible to ever "prove" that a log is "append-only" at all. I think the best thing you can prove is "all these statements (I am seeing) are part of the same log". It's definitely possible for a party to establish "I have always seen a single history", but whether that matches what's served to other parties require communication between them. Beyond consistency, there's also the question of whether the content was legitimately registered in the first place.

Different mechanisms are more or less useful depending on the nature of the actors interacting with the TS, and the operations they are willing to perform. In the end though "Append-only" is an aspiration, there's never a positive proof that a forked history will never turn up somewhere. There can only be proof that there definitively were forked histories, and there can be an absence of that for a long time, which is a very good sign.

Can we agree a term for "Sequence of Signed Statements" and use that where appropriate in conjunction with the VDS ("Sequence/Log/Ordered collection of Signed Statements committed to the VDS"?).

Reviewer 4: > Can we agree a term for "Sequence of Signed Statements" and use that where appropriate in conjunction with the VDS ("Sequence/Log/Ordered collection of Signed Statements committed to the VDS"?).

Yeah, I think that is the right angle here, no matter the position on provability.

Ideally, I would delegate all notions of provability to the VDS, which is the merkle proofs draft and its profiles. But do we want to make a concious choice on what is *required* of the VDS in this document ? I believe the current position is that the VDS is minimally required to prove inclusion and MAY provide other proofs such as append only/consistency. I am fine with that. 

*But* the discusion of sequence and timestamps in this discussion makes be uneasy about precisely where the arch stands on provability of those things: MUST/SHOULD/MAY ?

Reviewer 3: I am a little confused, because it seems to me that Receipts with an inclusion proof _are_ a commitment to a specific position, and that it _is_ possible for an auditor to verify that against a VDS/Sequence of Signed Statements if desired.
It is also true that consistency proof are an excellent optimisation to achieve the same result across a set of inclusion proofs, in cases where only consistency is desired, as opposed to verifying both order and correct/legitimate inclusion.

For that reason, I don't think being more restrictive would change security guarantees here, I suspect it runs the risk of overfitting to a specific use case. Similarly, the draft mentions distributed consensus and trusted execution environments as examples, but nothing more.

Reviewer 4: > I am a little confused, because it seems to me that Receipts with an inclusion proof _are_ a commitment to a specific position, and that it _is_ possible for an auditor to verify that against a VDS/Sequence of Signed Statements if desired. 

I don't think Receipts are required to commit the position in the current drafts. (Ours do though).

I believe what Receipts commit is currently defined by CoMETER and it doesn't specify this, it just uses RFC 9162 as an *example* of a conforming registration. I believe it is silent on notions of sequence and timestamps. And yet both seem important to the arch.

I also don't think Receipts are required to uniquely bind a signed statement to a location. It's entirely possible to register the same signed statement twice. At least in terms strictly of the proof material, both proofs would be verifiable and "correct", tho consumers may be confused and dispute the "sequence". 

And yes, a full audit shows the log is consistent with both receipts.

> It is also true that consistency proof are an excellent optimisation
Oh sure, I'm fine with the implication that "full audit" is the basic guarantee of worst case work.

Reviewer 4: to be super clear: if the answer here to my remark "But the discusion of sequence and timestamps in this discussion makes be uneasy" is "lets leave well enough alone" I am totally fine with this being the business of the VDS and the TS. I'm just surprised.

Reviewer 3: > I also don't think Receipts are required to uniquely bind a signed statement to a location. It's entirely possible to register the same signed statement twice.

True, and while opinions differ, I think they are separate registration events, potentially under different registration policies, will have different signatures, and therefore be different Receipts arguably.

> I don't think Receipts are required to commit the position in the current drafts. (Ours do though).

It's possible that there is a way to build a (secure) VDS/Inclusion VDP without including the position at all, but I don't know how it goes and I'd be curious to see it. You are right that CoMETER mentions RFC9162 as an example, and that does mention index-leaf in 5.2.
I think the position is important to decide what trust anchors and what registration policies are applicable for a given registration.

I am going to attempt a re-wording on the suggestion in the hope of landing somewhere that is compatible with our discussion :)

Reviewer 3: @robinbryce I think removing that is even better. If @SteveLasker is happy with, let's apply it?

Reviewer 6: This may be easier said than done. A registration authority will need to perform a "Legal Person" check in some cases. How would this activity be "replayed"?

Non-equivocation:

> : there is no fork in the Verifiable Data Structure.

Reviewer 3: ```suggestion
: there is no fork in the Append-only Log.
```

The VDS is there to make forks obvious, but the thing we are protecting is the log itself, not the tree.

Reviewer 5: A fork does not even make sense in the context of non linear VDS, so I agree that some change is needed here.

I think a better solution would be to frame "Non-equivocation" in terms of the VDS that does not assume a linear log... for example:

A property of certain implementations of VDS preventing conflicting statements between different processes.
In Byzantine fault tolerant systems, this property can be achieved with trusted hardware.

Author: If the existing PR or Suggestion isn't good enough, please make an alternative suggestion.

Reviewer 3: I have tried once more, with "sequence of Signed Statements", which is the log in my mind but is hopefully a more widely acceptable term. Does that work?

Author: Note: Without weighing in on the suggestion, I've updated the suggestion from TS --> Transparency Service.

Everyone with access to its content sees the same ordered collection of Signed Statements and can check that it is consistent with any Receipts they have verified.

Replayability:

> : the Verifiable Data Structure includes sufficient information to enable authorized actors with access to its content to check that each data structure representing each Signed Statement has been correctly registered.

Reviewer 3: ```suggestion
: the Append-only Log includes sufficient information to enable authorized actors with access to its content to check that each data structure representing each Signed Statement has been correctly registered.
```

The VDS only has digests, it does not contain enough information for an auditor to get roots of trust, registration policies etc. It's only enough to check consistency, not registration.

Author: Is this question going beyond the change from `Append-only Log` to `Verifiable Data Structure`?

Author: Calling @robinbryce for ⬆️

Reviewer 4: The MMR does not contain that information. My objection is to the implication that the TS must be responsible for making the information available to auditors. This is inherently centralizing and makes verifiabilty depend on the continued good operations of the TS.

A requirement that the TS must *define* how the information is available to auditors is fine by me. Its a one line change to make that work I think.

The move to replace "append only log" as a term with a "Registered Sequence of Signed Statements" helps a lot here I think - see https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/pull/351/files#r1975294713

Reviewer 6: I'm ok with ""Registered Sequence of Signed Statements""

This does not dictate implementation expectations the way that logs do.

Reviewer 6: +1

Author: Deferring to #348

The Transparency Service MAY validate the Signed Statement payload in order to enforce domain specific registration policies that apply to specific content types.
1. **Apply Registration Policy:** The Transparency Service MUST check the attributes required by a Registration Policy are present in the protected headers.
  Custom Signed Statements are evaluated given the current Transparency Service state and the entire Envelope, and may use information contained in the attributes of named policies.

> 1. **Register the Signed Statement** to the Verifiable Data Structure.

Reviewer 3: ```suggestion
1. **Register the Signed Statement** to the Append-only Log, and commit it to the Verifiable Data Structure.
```

Reviewer 5: Really don't like this.
To register a signed statement is to obtain a receipt from a VDS.... some VDS are append only log.
SCITT can assert that only certain VDS are suitable for this use case.

Author: Please provide an alternative suggestion?

Reviewer 7: Maybe just stop at "Register the Signed Statement"?

As Orie says above, the production of a receipt necessarily involves the use of one of the VDS options provided for in COSE Receipts, so removing words reduces redundancy, reduces opportunities for conflicts between references, and enables COSE Receipts to evolve usefully and independently.

Reviewer 6: +1 Maybe just stop at "Register the Signed Statement"

Reviewer 3: Amended the amendment accordingly.

1. **Return the Receipt**, which MAY be asynchronous from Registration.
The Transparency Service MUST be able to provide a Receipt for all registered Signed Statements.
Details about generating Receipts are described in {{Receipt}}.

> The last two steps may be shared between a batch of Signed Statements registered in the Verifiable Data Structure.

Reviewer 3: ```suggestion
The last two steps may be shared between a batch of Signed Statements registered in the Append-only Log.
```

Reviewer 5: Don't think we need this change

Reviewer 3: Agreed

Receipts are based on Signed Inclusion Proofs as described in COSE Receipts {{-COMETRE}} that also provides the COSE header parameter semantics for label TBD_0.

> The Registration time is recorded as the timestamp when the Transparency Service added the Signed Statement to its Verifiable Data Structure.

Reviewer 3: ```suggestion
The Registration time is recorded as the timestamp when the Transparency Service added the Signed Statement to its Append-only Log.
```

Reviewer 5: This would introduce the possibility of the VDS time and append only log time being different?
... also begs the question of "where is registration time" recorded.

Reviewer 4: Yes. we specicically included a record of time in the VDS so that it can be authorative. We do record a time at which the ts first recieves the registration request. But we see that firmly as part of the "adjacent service" data schema (not scitt). 

We ensure that "time of comitment to the vds" is consistent with the order of statements in the log and see this as very much the business of the TS and its VDS

OR remove the requirement for the registration time to be recorded altogether

Reviewer 3: CCF records it in the CWT_Claims.iat of the Receipt, because it seemed like the right place for it. I think what we are saying is that the VDS inclusion is what matters (that's where the proofs come from), and how the sequence of Signed Statements/log is stored is the business of an Adjacent service and comparatively less important, so it's not obvious that it deserves its own timestamp?

Reviewer 3: Following discussion, I agree this change is not helpful.

Transparency Services MAY support anonymous access.
Issuers SHOULD ensure Signed Statements submitted to public access services are acceptable for public disclosure.
Publicly accessible Signed Statements MUST NOT carry confidential information.

> Once a Signed Statement is inserted into the Verifiable Data Structure maintained by a Transparency Service, it cannot be removed.

Reviewer 3: ```suggestion
Once a Signed Statement is committed to the Verifiable Data Structure maintained by a Transparency Service, it cannot be removed.
```

1. Issuers can efficiently prove that their Statement is logged by a Transparency Service

The first guarantee is achieved by requiring Issuers to sign their Statements and associated metadata using a distributed public key infrastructure.
The second guarantee is achieved by storing the Signed Statement on an Verifiable Data Structure.

> The third guarantee is achieved by implementing a Verifiable Data Structure ({{-COMETRE}}).

Reviewer 3: That feels a bit redundant with the line above?

Issuers cannot be stopped from producing Signed Statements including false assertions in their Statement payload (either by mistake or by corruption), but these Issuers can made accountable by ensuring their Signed Statements are systematically registered at a Transparency Service.

Similarly, providing strong residual guarantees against faulty/corrupt Transparency Services is a SCITT design goal.
Preventing a Transparency Service from registering Signed Statements that do not meet its stated Registration Policy, or to issue Receipts that are not consistent with their Verifiable Data Structure is not possible.

> In contrast Transparency Services can be held accountable and they can be called out by any Auditor that replays their Verifiable Data Structure against any contested Receipt.

Reviewer 3: ```suggestion
In contrast Transparency Services can be held accountable and they can be called out by any Auditor that replays their Append-only log against any contested Receipt.
```

I don't think it's possible to do a full replay with just the digests, a replay from Signed Statements is necessary. As in other comments, I am perhaps being insufficiently imaginative when thinking about the VDS (which is a Merkle Tree in my mind).

Reviewer 4: This is actually suggesting "full audit" of the log material is the only way to hold the TS to account. Its true that this can be done, i don't think its true that this is the only or best way to accomplish this. 

```suggestion
In contrast Transparency Services can be held accountable and they can be called by Auditors for any contested Receipt. An Auditor in posession of a complete sequence of registered signed statements can perform a full audit. Auditors who have verified the state of the log prior to the Receipts issuance, can use consistency proofs from that state to more efficiently resolve the dispute.
```

Reviewer 3: Consistency proofs only work if what's disputed is consistency though, not correct registration checks. Should we separate the two kinds of audit?

Author: @achamayou, are you ok to commit the above change, and tweak? Or, do you have an alternative suggestion that we can commit?

Reviewer 3: I am not ok with the second change, because it claims that consistency proofs are sufficient to resolve a dispute. This is not true, they cannot confirm whether registration was performed correctly, and although they can be used to detect forks, they cannot be used to decide which fork(s) contain malicious content. I have amended my proposed change to remove the mention of "Append-only log" and use the term "Verifiable Data Structure", in an effort to find common ground.
